tests/test_auto_tokenize.py

```diff
@@ -0,0 +1,209 @@
+
+
+def test_nested_jits():
+    from mpi4jax import allreduce, auto_tokenize
+
+    arr = jnp.ones((3, 2))
+
+    @jax.jit
+    def f(a):
+        res, _ = allreduce(a, op=MPI.SUM)
+
+        @jax.jit
+        def g(b):
+            res, _ = allreduce(b, op=MPI.SUM)
+            res, _ = allreduce(res, op=MPI.SUM)
+            return res
+
+        res = g(res)
+        res = g(res)
+        return res
+
+    res = auto_tokenize(f)(arr)
+    np.testing.assert_allclose(res, arr * size ** 5)
+
+
+@pytest.mark.skipif(size < 2, reason="need 2 processes")
+def test_send_recv_tokenizer():
+    from mpi4jax import recv, send, auto_tokenize
+
+    def simple_message_pass(arr):
+        # Here, we test a crazy send/recv pattern that is extremely likely to deadlock unless
+        # the order is preserved.
+        if rank == 0:
+            b, _ = recv(arr, source=1, comm=comm)
+            return b
+        elif rank == 1:
+            a = arr + 1
+            send(a, dest=0, comm=comm)
+            return arr
+        return jnp.zeros_like(arr)
+
+    res = jax.jit(auto_tokenize(simple_message_pass))(jnp.zeros((2, 2)))
+    if rank == 0:
+        np.testing.assert_allclose(res, jnp.ones((2, 2)))
+    if rank == 1:
+        np.testing.assert_allclose(res, jnp.zeros((2, 2)))
+
+
+@pytest.mark.skipif(size < 2, reason="need 2 processes")
+def test_send_recv_hotpotato_tokenizer():
+    from mpi4jax import recv, send, auto_tokenize
+
+    def hot_potato(arr):
+        # Here, we test a crazy send/recv pattern that is extremely likely to return the
+        # wrong result unless the order is preserved.
+        if rank == 0:
+            a = arr + 1
+            b, _ = recv(arr, source=1, comm=comm)
+            send(a, dest=1, comm=comm)
+            c, _ = recv(arr, source=1, comm=comm)
+            d, _ = recv(arr, source=1, comm=comm)
+            e = c + d
+            send(e, dest=1, comm=comm)
+            f, _ = recv(arr, source=1, comm=comm)
+            send(e + f, dest=1, comm=comm)
+            send(e * d, dest=1, comm=comm)
+            return f
+        elif rank == 1:
+            a = arr + 2
+            send(a, dest=0, comm=comm)
+            b, _ = recv(arr, source=0, comm=comm)
+            send(a + b, dest=0, comm=comm)
+            send(a * b, dest=0, comm=comm)
+            c, _ = recv(arr, source=0, comm=comm)
+            send(b - c, dest=0, comm=comm)
+            d, _ = recv(arr, source=0, comm=comm)
+            e, _ = recv(arr, source=0, comm=comm)
+            return d + e
+        return jnp.zeros_like(arr)
+
+    # !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    # ! Proof that this test actually works !
+    # !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    #
+    # Uncomment the below line of code.
+    # The final result will be wrong and the asserts will fail.
+    #
+    # ---------------------------
+    # auto_tokenize = lambda x: x
+    # ---------------------------
+
+    real_result = hot_potato(jnp.zeros((2, 2)))
```

Comment:
Are there any guarantees that JAX doesn't execute anything out-of-order when not under JIT? I know it doesn't (at least not on CPU), but can we rely on this?

Reply:
Each operation is executed op-by-op when not using a jit, so I believe the order is always determined by the python execution order.

Reply:
Mmmh yea we probably won't get any better guarantees than that...

Reply:
I'm pretty sure that the GPU backend does not execute operations in order, judging from several deadlocks and segfaults I've experienced.

Reply:
That's what I feared, too. xla:cpu is by far the least aggressive optimizer of the bunch. I can try the hot potato and see what happens. If it reorders even non-JIT operations we would be in trouble, since there's no way to enforce proper execution order in something like this:

```python
autofunc = auto_tokenize(jit(f))

token = create_token()
arr, token = allreduce(...)
arr2 = autofunc(...)  # <- can't pass / return token
arr3, token = allreduce(...)
```

Reply:
Everything passes on (multi-)GPU. Looks like you're right.

Reply:
i'm glad to be proved wrong.